src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -78,20 +78,13 @@ private void StopRaisingEvents()
                 {
                     // Stop the RunLoop and wait for the thread to exit gracefully
                     Interop.RunLoop.CFRunLoopStop(_watcherRunLoop);
-                    _watcherThread.Join();
-                    _watcherRunLoop = IntPtr.Zero;
                 }
 
-                // Clean up the EventStream, if it exists
-                if (!_eventStream.IsInvalid)
-                {
-                    _eventStream = new SafeEventStreamHandle(IntPtr.Zero);
-                }             
-
-                // Cleanup the callback
-                if (_callback != null)
+                // Gurd against us from calling Join on our own background thread (for 
+                // example, calling Dispose or Stop from a callback).
+                if (Thread.CurrentThread != _watcherThread)
                 {
-                    _callback = null;
+                    _watcherThread.Join();
```

Comment:
That's true, that creates a bunch of complexity between the threads that makes it tricky to think and debug through. I'll close this PR and refactor the code to use the Cancellation Token like the Linux implementation....it shouldn't be too difficult and shouldn't take too long